### PR TITLE
OAK-9625 Support ordered index for first value of a multi-valued property, node name, and path

### PR DESCRIPTION
**Pull request by thomasmueller**

### CODE REVIEW

File: oak-core/src/main/java/org/apache/jackrabbit/oak/query/ast/FirstImpl.java [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+
+import org.apache.jackrabbit.oak.api.PropertyValue;
+import org.apache.jackrabbit.oak.api.Type;
+import org.apache.jackrabbit.oak.plugins.memory.PropertyValues;
+import org.apache.jackrabbit.oak.query.index.FilterImpl;
+import org.apache.jackrabbit.oak.spi.query.QueryConstants;
+import org.apache.jackrabbit.oak.spi.query.QueryIndex.OrderEntry;
+
+/**
+ * The function "first(..)".
```
Comment (joerghoh) [inline oak-core/src/main/java/org/apache/jackrabbit/oak/query/ast/FirstImpl.java:34]: Can you update the documentation for it as well?

File: oak-core/src/main/java/org/apache/jackrabbit/oak/query/ast/PathImpl.java [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+
+import org.apache.jackrabbit.oak.api.PropertyValue;
+import org.apache.jackrabbit.oak.api.Type;
+import org.apache.jackrabbit.oak.query.index.FilterImpl;
+import org.apache.jackrabbit.oak.plugins.memory.PropertyValues;
+import org.apache.jackrabbit.oak.spi.query.QueryConstants;
+import org.apache.jackrabbit.oak.spi.query.QueryIndex.OrderEntry;
+
+/**
+ * The function "path(..)".
```
Comment (joerghoh) [inline oak-core/src/main/java/org/apache/jackrabbit/oak/query/ast/PathImpl.java:35]: Can you update the documentation for it as well?

File: oak-core/src/main/java/org/apache/jackrabbit/oak/query/SQL2Parser.java [hunk 3]
```diff
@@ -676,6 +676,12 @@ private DynamicOperandImpl parseExpressionFunction(String functionName) throws P
             } else {
                 op = factory.nodeLocalName(readName());
             }
+        } else if ("PATH".equalsIgnoreCase(functionName)) {
```
Comment (fabriziofortino) [inline oak-core/src/main/java/org/apache/jackrabbit/oak/query/SQL2Parser.java:679]: I think you should add `PATH` and `FIRST` in the `getSyntaxError ` at line 707

File: oak-core/src/main/java/org/apache/jackrabbit/oak/query/ast/PathImpl.java [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+        }
+        return PropertyValues.newString(path);
+    }
+
+    @Override
+    public void restrict(FilterImpl f, Operator operator, PropertyValue v) {
+        if (v == null) {
+            return;
+        }
+        if (operator == Operator.NOT_EQUAL && v != null) {
```
Comment (fabriziofortino) [inline oak-core/src/main/java/org/apache/jackrabbit/oak/query/ast/PathImpl.java]: `v != null` is always true and can be removed

File: oak-core/src/main/java/org/apache/jackrabbit/oak/query/xpath/XPathToSQL2Converter.java [hunk 5]
```diff
@@ -731,6 +731,11 @@ private Expression parseFunction(String functionName) throws ParseException {
             f.params.add(parseExpression());
             read(")");
             return f;
+        } else if ("jcr:first".equals(functionName)) {
```
Comment (fabriziofortino) [inline oak-core/src/main/java/org/apache/jackrabbit/oak/query/xpath/XPathToSQL2Converter.java:734]: you should add these new cases in the `getSyntaxError` at line 808

**Review by fabriziofortino — CHANGES_REQUESTED** [review]

File: oak-core/src/main/java/org/apache/jackrabbit/oak/query/ast/FirstImpl.java (hunk 1 above)
Comment (thomasmueller) [reply oak-core/src/main/java/org/apache/jackrabbit/oak/query/ast/FirstImpl.java:34]: done

File: oak-core/src/main/java/org/apache/jackrabbit/oak/query/SQL2Parser.java (hunk 3 above)
Comment (thomasmueller) [reply oak-core/src/main/java/org/apache/jackrabbit/oak/query/SQL2Parser.java:679]: done

**Review by fabriziofortino — APPROVED** [review]

**Pull request merged**